lib/engine.js

```diff
@@ -0,0 +1,166 @@
+const { createStore, combineReducers } = require('redux');
+const _ = require('lodash');
+
+let store;
+
+// actions
+
+const PICK_OFFENSE_CARD = 'PICK_OFFENSE_CARD';
+const PICK_DEFENSE_CARD = 'PICK_DEFENSE_CARD';
+
+function pickOffenseCard(cardId) {
+  return {
+    type: PICK_OFFENSE_CARD,
+    cardId,
+    player: 'player',
+  };
+}
+
+function pickDefenseCard(cardId) {
+  return {
+    type: PICK_DEFENSE_CARD,
+    cardId,
+    player: 'player',
+  };
+}
+
+// reducers
+
+// game engine reducer helpers
+
+function validateCardChoice(choices, cardId) {
+  return choices && choices[cardId] ? choices[cardId] : null;
+}
+
+function removeFromDeck({ state, action, card }) {
+  const deckKey = action.player === 'ai' ? 'aiDeck' : 'playerDeck';
+  const cardInDeck = state[deckKey][action.cardId];
+  // if this is the last copy of the card, delete it from the deck
+  const updatedState = cardInDeck.count <= 1 ? {
+    ...state,
+    [deckKey]: _.omit(state[deckKey], action.cardId),
+  } : {
+    ...state,
+    [deckKey]: {
+      ...state[deckKey],
+      [action.cardId]: {
+        ...card,
+        count: card.count - 1,
+      },
+    },
+  };
+  return {
+    state: updatedState,
+    action,
+    card,
+  };
+}
+
+function playCard({ state, action, card }) {
+  // TODO: actually apply playing of a card
+  return { state, action, card };
+}
+
+function pickOpponentOffenseCard({ state, action, card }) {
+  // TODO: actually pick a card for opponent
```

Comment:
So if this is a helper in the _.flow for PICK_OFFENSE_CARD, and it is passed the player's picked offense card, what does it return for "card"?  that same card or the one picked for the opponent?  That's part of my confusion, is that we have one card object passed down through a flow that involves two cards being picked and acted on.